### ISSUE TITLE
Admin cannot add a new role to himself

### ISSUE THREAD

**Issue by mvarblow** [OrchardCMS/OrchardCore#16748]
### Describe the bug

Admin cannot add a new role to himself (or remove a role from himself) unless he is a site owner.

### Orchard Core version

2.0

### To Reproduce
1. Start with a fresh Orchard solution (empty app_data). Set up the site using the blog recipe and sqlite to keep it simple.
2. Go to '/admin' panel and sign in using the admin account (e.g. the original admin from site setup).
3. Edit the Administrator role, remove the "Site Owners Permission" permission, and click Save.
4. Create a new user named 'admin2', grant it the Administrator role, and copy the password.
5. Sign out, go back to the '/admin' panel, and sign in using the new admin2 username and password.
6. Go to the Users page, edit the admin2 user account, grant another role (e.g. Contributor), and click Save.
7. See the success notification. 
8. Edit your admin user account and see that the notification was a lie. Your user still has the same roles as before.
9. Scratch your head and wonder what just happened.
10. Curse and repeat step 6.

### Expected behavior
Your user account should be updated to include the roles you selected (or to remove the ones you deselected) when you edited it. If the current behavior is as documented, I couldn't find the documentation. And it is incredibly confusing for the admin.

Editing your own roles is allowed for a site admin. I don't understand (and our clients don't understand) why this should not be allowed for other admins who are granted the permission to edit user role assignments. A safeguard against getting locked out of the site? I could understand that. But then why is it allowed for site owners? And isn't that why OrchardCore includes this check? 

![image](https://github.com/user-attachments/assets/ed696883-2940-40f2-b5c0-0572f6ef6e3b)

It seems that at a minimum, there should be a notification that the change of roles was blocked. The success message in this case is very confusing and I think makes this a bug. However, I think this check should be removed. It's confusing. It doesn't seem justified given that we already have user permissions to control who can change user roles. And there is already a sanity check to ensure that the only site admin hasn't removed their admin rights. If we wanted to do something like this, it should be done with a distinct permission (e.g. "Edit my own roles") not with the Site Owner permission.

(Possibly related to https://github.com/OrchardCMS/OrchardCore/issues/8977. Also, it looks like there used to be a notification that you couldn't change your own roles but it was removed because it was appearing in a front-end user profile edit endpoint? I suspect it was unintentionally removed for this use case where the user is edited from the admin panel?)

**Comment by MikeAlhayek** [comment]
I recall fixing something like that at one point. From the code you shared, we are only guarding against locking yourself out. The `SiteOwner` is only check to ensure only site owners can change roles.

I do agree this condition is wrong.
https://github.com/OrchardCMS/OrchardCore/blob/5caef1c251dd9069240573e71debcb78b18c01bd/src/OrchardCore.Modules/OrchardCore.Users/Drivers/UserRoleDisplayDriver.cs#L94-L99 because we later preventing an admin from removing their admin role of they are the only active user.


Instead of `StandardPermissions.SiteOwner` it should be `CommonPermissions.AssignRoleToUsers`

**Comment by mvarblow** [comment]
@MikeAlhayek I ran several scenarios through the debugger. I think it is actually safe, and preferable, to just remove that check entirely. And also remove the check from the Edit method. Those restrictions are not necessary. The Edit and UpdateAsync method both already have logic in place to limit the user to assigning (or even viewing assignments) to roles that the user has permission to assign to.

```
        var roles = await GetRoleAsync();
        // Authorize each role in the model to prevent html injection.
        var accessibleRoleNames = await GetAccessibleRoleNamesAsync(roles);
        var currentUserRoleNames = await _userRoleStore.GetRolesAsync(user, default);

        var selectedRoleNames = model.Roles.Where(x => x.IsSelected).Select(x => x.Role);
        var selectedRoles = roles.Where(x => selectedRoleNames.Contains(x.RoleName, StringComparer.OrdinalIgnoreCase));
        var accessibleAndSelectedRoleNames = await GetAccessibleRoleNamesAsync(selectedRoles);

```

I'm preparing a pull request.

**Comment by MikeAlhayek** [comment]
Its better to check for `AssignRoleToUsers` early on so we are not doing the rest of the logic if you can't assign roles.

**Comment by mvarblow** [comment]
@MikeAlhayek, wouldn't that break the page for users who are allowed to assign some, but not all roles? I think the existing logic does this correctly. Anything we put in front of it is more likely to break it.

**Comment by MikeAlhayek** [comment]
@mvarblow okay. I did more looking at the code. I agree with you. I also notices a descrepency that should be addressed "could be in this PR or another PR".

There are two permission related to assigning role to users 

https://github.com/OrchardCMS/OrchardCore/blob/dd03cf8c998b014303728e20b398239e49ff3dc8/src/OrchardCore/OrchardCore.Roles.Core/CommonPermissions.cs#L9

and 

https://github.com/OrchardCMS/OrchardCore/blob/dd03cf8c998b014303728e20b398239e49ff3dc8/src/OrchardCore/OrchardCore.Users.Core/CommonPermissions.cs#L24

One of these two should be removed

The one that SHOULD be kept in this one 
https://github.com/OrchardCMS/OrchardCore/blob/dd03cf8c998b014303728e20b398239e49ff3dc8/src/OrchardCore/OrchardCore.Users.Core/CommonPermissions.cs#L24

We can add a migration step that would update all existing roles that contains "ManageRoles" or "AssignRoles" claim, and auto assign "AssignRoleToUsers" just to prevent a breaking change.

**Comment by MikeAlhayek** [comment]
@mvarblow actually, don't bother with removing the AssignRoles. I think it is more complicated and should be in a separate PR. But please proceed with your PR. ping me when it's done and I'll review it.

**Comment by sebastienros** [comment]
"Admin cannot add a new role to himself (or remove a role from himself) unless he is a site owner"

I have seen that kind of wording in too many places, convoluting "admin" with the `SiteOwner` permission. It's totally fine to not have this permission on an Administrator role. But then we should not expect users to have "SiteOwner" powers because they are in a role named "Admin". Or keep the Administrator role with `SiteOwner` permission, but don't assign it to users that easily.

So let me rephrase: "Random users can't add or remove new role to/from themselves unless they have the SiteOwner permission"

By design, see https://github.com/OrchardCMS/OrchardCore/issues/8977#issuecomment-808045750

If the issue is about fixing the UI to explain why, or showing the correct notification, I find it premature to just say that we should remove these checks.

I may definitely be wrong and change my mind in the next hour, but after spending some time on this issue this is my current feeling. I'll come back with more comments once I try that myself ;)

**Comment by mvarblow** [comment]
@sebastienros 

> By design, see https://github.com/OrchardCMS/OrchardCore/issues/8977#issuecomment-808045750

The comment you reference says:

> This was done to prevent a user removing themselves from all roles, and then being unable to access the site.

Why are we concerned that a non-SiteOwner admin can remove themselves from the Administrator role but we're not concerned that a SiteOwner can remove themselves from the Administrator role. There is already a separate check to prevent the last (i.e. there is only one) Administrator from removing themselves from the Administrator role. And there are already checks to ensure that the user has permission to make changes to the specific role assignments. I.e. there are checks to be sure that a user cannot remove the X role from any user unless they have been granted permission to manage X role assignments. 

This additional check that a non-SiteOwner cannot manage their own roles doesn't seem to enforce any rules that are described in the OrchardCore documentation. And it seems unnecessary/redundant given the permissions, and sanity check on removing the Administrator role from the last Administrator, that are already in place.

**Comment by mvarblow** [comment]
Also, @sebastienros, it looks like you removed this from 2.0.2? If that's correct, it would be helpful to update the release notes. They seem to indicate now that it would be included -- it's listed as *intentionally* included in 2.0.1 and the 2.0.2 notes say that 2.0.2 removed commits that were *not* intentionally included.

**Comment by sebastienros** [comment]
> it looks like you removed this from 2.0.2

We realized there was an unrelated issue with 2.0.1, and while redoing the release we decided we should not include this change since it's not "patch" worthy and there is a security note to it.

I will update the release notes, thanks.
